test/distributed/test_c10d_nccl.py

```diff
@@ -968,6 +969,51 @@ def test_send_recv(self):
                 dist.send(send_tensor_view, 1)
 
 
+    @unittest.skipIf(TEST_WITH_ROCM or
+                     int(torch.version.cuda.split(".")[0]) < 11, "CUDA >= 11.0 required for graphs")
+    @skip_if_lt_x_gpu(2)
+    def test_sync_batch_norm_cuda_graph_capture(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        pg = self._create_process_group_nccl(store, self.opts())
+
+        model = torch.nn.Sequential(
+            nn.BatchNorm2d(2),
+            nn.Linear(30, 2),
+        ).to(device=self.rank)
+
+        model = nn.SyncBatchNorm.convert_sync_batchnorm(
+            model,
+            process_group=pg,
+        )
+
+        optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
+
+        model.train()
+
+        static_input = torch.zeros(
+            (3, 2, 30, 30),
+            dtype=torch.float32,
+            device=self.rank
+        )
+
+        # capture
+        g = torch.cuda.CUDAGraph()
+        with torch.cuda.graph(g):
+            model(static_input).sum().backward()
```

Comment:
Hey @mcarilli, I am testing SyncBatchNorm with cuda graph capturing. Looks like allgather is not supported either? But the original code also uses allgather, and I saw your tests have passed allgather. Did I miss anything here?

```
a100-st-p4d24xlarge-7:27858:27858 [0] init.cc:928 NCCL WARN Cuda failure 'operation not permitted when stream is capturing'                                     
a100-st-p4d24xlarge-7:27858:27858 [0] NCCL INFO init.cc:953 -> 1                                                                                                
ERROR:torch.testing._internal.common_distributed:Caught exception:                                                                                              
Traceback (most recent call last):                                                                                                                              
  File "/fsx/users/shenli/pytorch/test/distributed/test_c10d_nccl.py", line 1002, in test_sync_batch_norm_cuda_graph_capture                                    
    model(static_input).sum().backward()                                                                                                                        
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1129, in _call_impl                                
    return forward_call(*input, **kwargs)                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/container.py", line 139, in forward                                 
    input = module(input)                                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1129, in _call_impl                                
    return forward_call(*input, **kwargs)                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/batchnorm.py", line 748, in forward                                 
    return sync_batch_norm.apply(                                                                                                                               
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/_functions.py", line 56, in forward                                 
    dist._all_gather_base(combined_flat, combined, process_group, async_op=False)                                                                               
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 2136, in _all_gather_base               
    work = group._allgather_base(output_tensor, input_tensor)                                                                                                   
RuntimeError: NCCL error in: /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126, unhandled cuda error, NCCL version 2.10.3                 
ncclUnhandledCudaError: Call to CUDA function failed.   
```

Reply:
Do I need to update to the latest nccl 2.12.12-1?

Reply:
Tested nccl 2.12.12-1, same error

```
1.13.0a0+git94153e1                                                                                            
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 1                                                               
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 0             
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.                                   
NCCL version 2.12.12+cuda11.1                                                                                                                 
                                                                                                                                                                
a100-st-p4d24xlarge-7:7834:7834 [0] init.cc:938 NCCL WARN Cuda failure 'operation not permitted when stream is capturing'                                       
a100-st-p4d24xlarge-7:7834:7834 [0] NCCL INFO init.cc:963 -> 1                                                               
ERROR:torch.testing._internal.common_distributed:Caught exception:                                                
Traceback (most recent call last):                                                                                                 
  File "/fsx/users/shenli/pytorch/test/distributed/test_c10d_nccl.py", line 1004, in test_sync_batch_norm_cuda_graph_capture                                    
    model(static_input).sum().backward()                                                                                                                        
  File "/fsx/users/shenli/pytorch/torch/nn/modules/module.py", line 1130, in _call_impl                           
    return forward_call(*input, **kwargs)                                                                      
  File "/fsx/users/shenli/pytorch/torch/nn/modules/container.py", line 139, in forward                        
    input = module(input)                                                                                      
  File "/fsx/users/shenli/pytorch/torch/nn/modules/module.py", line 1130, in _call_impl                           
    return forward_call(*input, **kwargs)                                                                         
  File "/fsx/users/shenli/pytorch/torch/nn/modules/batchnorm.py", line 748, in forward                            
    return sync_batch_norm.apply(                                                                                                                               
  File "/fsx/users/shenli/pytorch/torch/nn/modules/_functions.py", line 56, in forward                                             
    dist._all_gather_base(combined_flat, combined, process_group, async_op=False)                                                                                 File "/fsx/users/shenli/pytorch/torch/distributed/distributed_c10d.py", line 2136, in _all_gather_base          
    work = group._allgather_base(output_tensor, input_tensor)                                                     
RuntimeError: NCCL error in: /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126, unhandled cuda error, NCCL version 2.12.12
ncclUnhandledCudaError: Call to CUDA function failed.                                                         
Exception raised from create at /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126 (most recent call first):                               frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7ff141fc29db in /fsx
/users/shenli/pytorch/torch/lib/libc10.so)      
```

Reply:
Try
```python
os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "0"
torch.distributed.init_process_group(...)
```
following https://pytorch.org/docs/master/notes/cuda.html#usage-with-distributeddataparallel. That section title really should be "Usage With NCCL"...

Reply:
Same error, curious, why using `NCCL_ASYNC_ERROR_HANDLING` might make a difference here?

Reply:
The timeout checking incurs some uncapturable cuda calls (event queries i think, i'd have to double check the code)